### ISSUE TITLE
Add a custom-nat chain in the firewall

### ISSUE THREAD

**Issue by rapenne-s** [QubesOS/qubes-issues#8629]
### The problem you're addressing (if any)

With the move to nftable, there are chains like `custom-input` and `custom-forward`. From my experience when writing a community guide for nftables https://forum.qubes-os.org/t/qubes-os-4-2-nftables-nft-firewall-guide/20933 , we need `custom-nat` for the NAT rules

### The solution you'd like

A `custom-nat` chain to add nat rules.

### The value to a user, and who that user might be

People could add nat rules without having to create a chain

**Comment by 3hhh** [comment]
If you just want to overwrite the Qubes OS NAT decision, this works just fine:
https://github.com/3hhh/qubes-dns/blob/8b2a195c43647982ed6d08a72c0f29262c829cad/dns_client#L80-L81
If you want to let Qubes OS decide first, use `+1` instead.
That method is even backwards compatible to Qubes OS 4.1, which any new method wouldn't be.

So in total I'm not sure whether `custom-nat` is really needed.

**Comment by rapenne-s** [comment]
> If you just want to overwrite the Qubes OS NAT decision, this works just fine: https://github.com/3hhh/qubes-dns/blob/8b2a195c43647982ed6d08a72c0f29262c829cad/dns_client#L80-L81 If you want to let Qubes OS decide first, use `+1` instead. That method is even backwards compatible to Qubes OS 4.1, which any new method wouldn't be.

Do you mean that any extra NAT table should be done with a lower priority? I'm not sure I understand the `+1`, do you mean one should add 1 to the current priority, or litteraly use `+1` (I can't find anything like +1 in the doc so I'd be sure I'm not missing something ;) )

> 
> So in total I'm not sure whether `custom-nat` is really needed.

I'm not sure I totally follow here. I think it's better to not touch the defaults rules, this is certainly why there are `custom-forward` and `custom-input` instead of modifying existing chains. This would make sense to have a `custom-nat` for consistency.

**Comment by 1cho1ce** [comment]
Some thoughts about how to add custom rules in a better way and links about nft chain priority here:
https://github.com/QubesOS/qubes-issues/issues/8562#issuecomment-1762682079

**Comment by 3hhh** [comment]
On 10/24/23 09:50, 1cho1ce wrote:
> Some thoughts about how to add custom rules in a better way and links about nft chain priority here:
> https://github.com/QubesOS/qubes-issues/issues/8562#issuecomment-1762682079

IIRC it doesn't work that way for filter tables as the default policy of Qubes OS is to deny and that would be a final decision regardless of the decision you made before or are attempting to make afterwards (your callback will not be called anymore by nftables once a deny was encountered). So custom chains for filter tables are required.

Since the Qubes OS NAT policy is "allow" by default, it matters less there unless Qubes OS starts to deny stuff on NAT level somehow.

Solene: My point was that `custom-nat` is technically not required. It may still make sense for aesthetic reasons, idk. The explanations can be found in the comment above and my code reference.

**Comment by rapenne-s** [comment]
> My point was that `custom-nat` is technically not required. It may still make sense for aesthetic reasons, idk. The explanations can be found in the comment above and my code reference.

this is true for custom-input and custom-forward right? ;) Nothing prevents users from injecting the rules at the right place, but yet these chains exist too

**Comment by rapenne-s** [comment]
what command line would you recommend to Qubes OS users for a NAT rule then? The dns_client example is also creating a new NAT chain.

**Comment by 1cho1ce** [comment]
> IIRC it doesn't work that way for filter tables as the default policy of Qubes OS is to deny and that would be a final decision regardless of the decision you made before or are attempting to make afterwards (your callback will not be called anymore by nftables once a deny was encountered). So custom chains for filter tables are required.

Yes, you're right, it won't work that way for custom chains with drop policy (input chain):
https://wiki.nftables.org/wiki-nftables/index.php/Configuring_chains#Base_chain_priority

So I guess the better way would be to create a separate custom chain for separate things and just add the jump to this separate chain to default custom-X qubes chain?
For example, you have qube with VPN client that want to add/remove its custom firewall rules on start/stop and some server is running in this qube that needs to add firewall rules to be accessed from other qubes, connected to this qube.
If you just add all the rules in a single default custom chain then it'll be hard to remove them later.
Then, using input hook as an example, in firewall up/down script you need to check if its dedicated custom chain already exist and if not then create this chain and add jump to it in `custom-input` chain:
```
if nft create chain ip qubes custom-myvpn-input '{ type filter hook input priority filter; policy accept; }') &>/dev/null; then
    nft add rule ip qubes custom-input jump custom-myvpn-input
fi
```
And then use `custom-myvpn-input` to add or flush its custom rules.
The same goes for the server in this qube and its service should check the same thing on start.

**Comment by marmarek** [comment]
> > My point was that `custom-nat` is technically not required. It may still make sense for aesthetic reasons, idk. The explanations can be found in the comment above and my code reference.

Exactly.

> this is true for custom-input and custom-forward right? ;) Nothing prevents users from injecting the rules at the right place, but yet these chains exist too

In case of input/forward, you need to inject rules into some of the already defined chains, especially if you want to allow something, otherwise "deny" default rule in them would reject the traffic anyway. custom-input/custom-forward chains are convenient place for it. Strictly speaking, it isn't necessary in other chains because qubes firewall doesn't place default "deny" in them.

**Comment by Eric678** [comment]
My 2c: definitely worth having custom-* chains as you never know when the devs are going to change something and you get an unpleasant surprise after an upgrade.  I created a custom-dnat chain everywhere.

**Comment by rapenne-s** [comment]
> My 2c: definitely worth having custom-* chains as you never know when the devs are going to change something and you get an unpleasant surprise after an upgrade. I created a custom-dnat chain everywhere.

I think we can recommend people to create a `custom-$qube-dnat` chain when they need dnat, this ease flushing rules. But this loses consistency with the two other custom chains
